Add user passwords (#301)

inventor02 opened the pull request:
Fixes #301

This PR adds passwords to users and if the right config values are set, prompts them for their password when they use a token they have not used for a while. If configured, users will be forced to set a password upon login if they do not have one already.

It also allows users to login directly with their user ID and password, if the configuration option `user:allow_password_login` is enabled. To support this, in any case, the user ID is surfaced a few more times around the UI, namely in the edit user dialogs and on the current user information dialog.

Passwords are PBKDF2 hashed + salted with the necessary information to check them stored in the database. This means alternative hashing methods or implementations could be added without breaking previously-set passwords.

Passwords can be changed by anyone with the `edit-user-password` permission (`manager`s by default). Users can change their own passwords, but only if they have the `edit-current-user-password` permission (which `basic_user`s do by default).

![image](https://github.com/user-attachments/assets/cc7dccca-dd8f-4670-a55a-ba46d9c2135d)

![image](https://github.com/user-attachments/assets/a636cec3-8cd9-466a-936f-526aa7bb42c9)

![image](https://github.com/user-attachments/assets/947498a3-be7b-4a34-8c26-e4d9a4224738)

![image](https://github.com/user-attachments/assets/40088a53-cbcc-4b68-b458-a0857f4d5421)

![image](https://github.com/user-attachments/assets/53a537dd-f6ec-4569-a195-ae81149ffbb0)

![image](https://github.com/user-attachments/assets/af815fa3-b2ab-431a-8180-b9d4a98de87c)

![image](https://github.com/user-attachments/assets/e6f05248-17b8-4bcb-a5d9-c3bf61d5c748)

inventor02 committed:
add password fields to schema

inventor02 committed:
add password related config options

inventor02 committed:
add ability to set password

inventor02 committed:
add hidden fields

inventor02 committed:
improve docs

inventor02 committed:
password prompt at login

inventor02 committed:
fix hidden fields

inventor02 committed:
rearchitect login to use callbacks

inventor02 committed:
add password check to stockterminal

inventor02 committed:
refactor function nesting

inventor02 committed:
update token last login

inventor02 committed:
dodgy hack to only display the prompt once

inventor02 committed:
update tillweb

sde1000 commented:
Hi! Thank you for looking at this. This implementation looks good to me, at least at first glance...

I think the options you currently set via the command line and tillconfig module would be better off as `quicktill.config` options — probably declared in the `quicktill.user` module? The reasoning being, command-line options are for things that affect particular instances of the till software, whereas `quicktill.config` options affect the entire installation, and I think the password feature needs to be configured for the installation as a whole.

Open questions:

* How should passwords be stored? Plaintext? You're using sha256 of the password, which is reasonable, but would it be better to use an industry-standard password hash instead?
* I'm a bit iffy about the global `user.is_password_prompt_displayed` — I suspect it's covering up an underling problem...

inventor02 commented:
> I think the options you currently set via the command line and tillconfig module would be better off as quicktill.config options

Initially I thought it might be desired to configure stuff like this separately for different types of terminal but thinking about it now, yes, I agree.

> How should passwords be stored? Plaintext? You're using sha256 of the password, which is reasonable, but would it be better to use an industry-standard password hash instead?

When I was testing this I was storing them in plain text - the idea I had was that however they're stored, the function between passwords and storage mechanisms should be deterministic so that it is possible to identify a user by their password (e.g. if they forget a token, or if passwords are not issued at all, they can log in with a password alone, if enabled).

We could of course also store them in plain text - which sounds bad, but I think with adequate warnings to the user could be doable...

> I'm a bit iffy about the global user.is_password_prompt_displayed — I suspect it's covering up an underlying problem...

It is - the issue is that because tokens are handled at a very low level, if tokens are presented multiple times then multiple of the password entry dialogs are shown. I need to look closer at this and work out a nicer way of solving this problem (or if you have any immediate thoughts that'd be helpful too! :-))

inventor02 committed:
change configs to configitems

inventor02 committed:
lint user.py

inventor02 committed:
lint ui.py

inventor02 committed:
lint stockterminal.py

inventor02 committed:
lint register.py

inventor02 committed:
handle hotkeypress() in password prompt

inventor02 committed:
don't double-wrap users!

inventor02 committed:
don't use nested functions here

inventor02 committed:
implement comparison operators for IntConfigItem

inventor02 commented:
I stumbled across ui.hotkeypress() which when implemented allows us to catch usertoken events, and dismiss the dialog

inventor02 committed:
change operators implemented

inventor02 committed:
add dismissable param

inventor02 committed:
force setting password on login if required

inventor02 committed:
improve message formatting

inventor02 committed:
change colour of info dialog

inventor02 committed:
change cleartext

inventor02 committed:
add password-only logon and unique password requirement

inventor02 committed:
lint user.py

inventor02 committed:
pass user to stockterminal

inventor02 committed:
fix config item handling (oops)

inventor02 committed:
dont allow multiple logon dialogs

inventor02 committed:
dont force LOGON button

inventor02 committed:
oops...

inventor02 committed:
update last seen when logging in with password

sde1000 commented:
Hi! I'm sorry it's taken me so long to get around to looking at this. I think it's getting close to ready.

# Password-only login

I hadn't thought about implementing this! On reflection I think it is a good idea.

However, I don't think it should be possible to look a user up by their password alone — see the discussion below about how to store passwords.

We already have the concept of a "user ID" — it's just the primary key of the `User` model. It isn't exposed in many places at the moment; currently it's just shown in the output of the command-line `listusers` command, and accepted by the `--user` command-line option. We could show this in a few more places (in the users menu, the user info popup and the edit-user box) and then prompt for both user ID and password when people log in without a usertoken.

# Storing passwords

Since these passwords are only going to be used for local access to the till, not for remote login to the web interface, I suspect they will mostly be quite short and numeric — four or five digit numbers, most likely. (Some tills don't have an alphanumeric keyboard, so the number pad is all you have for password entry.)

People being people, I think it is quite likely that some of them will re-use PINs from other places, like payment cards. It would be irresponsible to store them in plaintext!

I think the only question the database should be able to answer is of the form "does user x have password y" so passwords can be checked at login time. It shouldn't be able to answer the question "what is user x's password?", and it shouldn't be able to answer the question "do any users have password y" without going through and checking all of them explicitly. (I.e. there should be no index from password to user, the config option `require_unique_passwords` should be removed, and two users having the same password should *not* have the same value stored in the password field.)

I think the best thing to do would be to take inspiration from Django — https://docs.djangoproject.com/en/5.1/topics/auth/passwords/

We don't necessarily have to use PBKDF2, but the idea of having the algorithm, salt, and work factor be stored in the database alongside the hash is a good one.

The "check if a password is correct for this user" operation should probably be a method on the `User` model.

# Assigning user tokens

When a user token is assigned to a new user (in `user.manage_user_tokens` or `user.managetokens`) its `last_successful_login` should be set to null.

inventor02 committed:
Merge branch 'sde1000:main' into user-pins

inventor02 committed:
show user ID in edit dialogs

inventor02 committed:
change allow_password_login name and description

inventor02 committed:
show ID in current user info

inventor02 committed:
get UID at login

inventor02 committed:
clear on TAB

inventor02 committed:
improve error message

inventor02 committed:
switch to pbkdf2

inventor02 committed:
clear last_successful_login when tokens are (re)assigned

inventor02 committed:
handle navigation away from UID

inventor02 committed:
remove unique passwords config setting

inventor02 commented:
Don't worry about it - it's been a very busy time of year for me too :-)

I think I've addressed all your feedback? Please have a look/play and let me know what you think.

I agree about adding the UID to the login flow for password-only authentication - I've amended the PR to do this. The UID is shown in various places now, and when logging in with the password-only key, the UID is entered first (and then checked, and the shortname is shown next to it).

The password hashes are now PBKDF2 (Django-style tuple storage in the database) too.

I don't *think* I missed anything but it's entirely possible so let me know if you spot anything, or find any of it confusing/hard to use!

sde1000 commented:
Just a few notes now...

In `localutils.py` you've changed the stockterminal hotkey to create a stock terminal page with the current user, rather than with no user. Although that's valid and probably useful behaviour, it will be confusing to existing till users who expect the stock terminal page to have no permissions. I don't think it's connected to the "user-pins" feature so it should be left unchanged here.

You've added `ui.validate_positive_int()` for the user ID field, but user IDs start at 1 so the existing `ui.validate_positive_nonzero_int()` would work better for that.

Something about `user.token_password_timeout` is bothering me. I don't think having different behaviour for "0" is useful — in practice it would be the same as "1". I think it should be a `config.IntervalConfigItem` where blank or zero interval disables checks. Perhaps it should be named `user.password_check_after` to be consistent with things like `register.transaction_lock_after`?

The `user.require_unique_passwords` config setting still needs removing.

I'm undecided about whether the password hash generation and checking code should be part of the `models.User` class, or whether it should be in a separate module. It shouldn't be in `user.py`, because it needs to be usable by the web interface code as well so we can add the "reset password" facility to the web interface. Maybe create a `passwords.py` module for this code?

In `user.edituser`, the user ID doesn't need to be in an editfield; it's immutable, so it can just be drawn as text when setting up the popup.

sde1000 commented:
Also, I've just realised: if `user.force_password_registration` is set to False, there should be an option in the UI for users to remove their passwords too.

sde1000 commented:
One more thing that has just become an issue since I checked in the sqlalchemy-2.0 migration code: all your uses of `td.s.query(model).get(pk)` in `user.py` need changing to `td.s.get(model, pk)` to avoid triggering sqlalchemy deprecation warnings during testing.

inventor02 committed:
do not pass user to stockterminal

inventor02 committed:
use validate_positive_nonzero_int for UID

inventor02 committed:
refactor: move password hash and check into new module

inventor02 committed:
change token_password_timeout to interval and rename

inventor02 committed:
change sqlalchemy get to 2.0-compatible calls

inventor02 committed:
allow users to remove their own passwords

inventor02 committed:
style: add newline at the end of passwords.py

inventor02 committed:
draw user ID as a string in the edit user dialog

inventor02 committed:
Merge branch 'sde1000:main' into user-pins

inventor02 committed:
use last_successful_login instead of last_seen

inventor02 committed:
use last_seen, but only set it after successful logins

inventor02 committed:
fix config item name

sde1000 commented:
> I've changed it so if force_password_registration is disabled, blank values are accepted and clear the password. Do you think people with the edit-user-password permission (managers) should always be able to clear user passwords, irrespective of the force_user_passwords setting? It could be useful, as if a user forgets their password they could just have it cleared, and then set their own the next time they log in to the till.

Yes, I think people with permission to edit other users' passwords should be able to set them to blank, so the user is forced to set a new password next time they visit the till.

I think the logic around when to check the password still needs some thought.

For example, starting with a till database with no passwords set at all I did the following:

* Set user:password_check_after to "5 minutes" but left the other two configuration options at their defaults of "No"
* Started the till and touched in using a red token — this sets `UserToken.last_seen` but does not alter `UserToken.last_successful_login` (so it's still null)
* Set my own password to "1234"
* Locked the till
* Used the red token again — this logged me in with no password prompt and did not set `UserToken.last_successful_login`, which is not what I was expecting!

We have three timestamp fields to think about: `User.last_seen`, `UserToken.last_seen` and `UserToken.last_successful_login`. I think these should be updated as follows:

* `User.last_seen`: when the user successfully accesses the till, with or without the use of a password
* `UserToken.last_seen`: whenever the token is presented, successfully or otherwise
* `UserToken.last_successful_login`: when the user has a password set, and presentation of the token leads to the user successfully accessing the till (with or without being prompted for the password), and not at any other time (apart from being set to null when the token is reassigned)

`User.last_seen` is currently updated in `user.user_from_token()`, so it'll be updated even when a password is prompted for and not provided — this needs to change.

I think `user.user_from_token()` should be removed, and its functionality should be split across `user.token_login()` and maybe a new function `user._finish_login()` that takes care of updating `User.last_seen` and invoking the callback to the register or stock terminal — this would be called directly by `user.token_login()` if no prompt is needed, or by `user.password_prompt()` and `user.password_login_prompt()` if a password is being supplied.

The check in `user.should_prompt_for_password()` is looking at `UserToken.last_seen` and ignoring `UserToken.last_successful_login` — this needs to be the other way around.

It's generally not necessary to call `td.s.commit()` in most of the till code — alterations to the ORM session will be committed automatically once the current keypress has finished being handled. If you have a search for it you'll see it's really not used very much — only in the `squareterminal.py` and `xero.py` modules where we genuinely do need to commit to the database and then continue doing more database operations because we are interacting with external services over the network. The existing uses in `barcode.py` are down to me not thinking clearly at the time and I really ought to remove them!

inventor02 committed:
refactor user.py

inventor02 committed:
always allow clearing another user's password if you have permission

inventor02 committed:
remove duplicate check

inventor02 commented:
That makes sense - sorry, life and things sort of happened between starting this work and coming back to look at it in the new year and so it has taken me a while to get back up to speed with what I actually did!

I have refactored user.py and removed `user_from_token()` as suggested. The token information is fetched in `token_login()` instead. All the login functions now finalise with `_finish_login(dbuser, callback)` which then passes the (wrapped) user to the till or stock terminal, and updates the `last_seen` attribute in the `users` relation.

The `usertokens.last_successful_login` attribute is now set when the password is checked and is correct, and when the password is not prompted for, but the user *does* have a password set. `usertokens.last_seen` is always set whenever a token is presented and it is a valid one to log in with (i.e. it has a user, and the user is enabled). I am not sure about this behaviour - I wonder if it makes sense? Or do we care about tracking when the token was last seen at all, even if it is not assigned to a user?

I've tested this in a few scenarios and combinations of settings and it appears to work - although I don't actually own anything that can read tokens (despite having about n+1 iButtons and RFID things...) so I am limited to debug -> fake token.

Also, I removed all the instances of `td.s.commit()` I added - that one's down to me not being familiar with SQLAlchemy!

sde1000 commented:
I got a "force set password" prompt in testing, despite `user:force_password_registration` being set to "No".

I think the logic for this would be clearer if it was all implemented in `user.token_login()` — at the moment you call `password_prompt` if either of `should_prompt_for_password()` or `should_force_set_password()` is True, and then `password_prompt` ignores the state of `should_force_set_password()` and falls through to `change_current_user_password_login_initial()`.

I think `password_prompt` should fall through to `_finish_login()` if the user doesn't have a password set, and the decision to call `change_current_user_password_login_initial()` should be made in `token_login()`.

You dismiss the password prompt popup if a new usertoken is presented before the password is entered, but you don't dismiss any of the others like the password setting popup and the prompt for setting a new password. I was able to stack these arbitrarily deep by presenting usertokens in sequence with no password set!

I don't think the popups for setting a new password at login should be un-dismissable — the user should always be able to back out with the Clear key (although they would end up not logging in). Having a box stuck on the screen will make users panic and say "the till is broken".

When checking the password, it would be nice if the user didn't have to press Cash/Enter after entering a correct password. You could use the same technique as `stock.stockpicker` to check the password every time the password field contents change. (Still watch for the user pressing Cash/Enter, though, so you can say "wrong password" explicitly.)

sde1000 commented:
(Maybe make a mixin class for "dismiss this popup when a usertoken is presented or the LOGIN key is pressed", with the `hotkeypress()` method you currently use in `password_prompt`?)

sde1000 commented:
I've just made what I hope is the final release in the v23 series. If it's ok with you, I'll merge this pull request, fix up the remaining issues, fix some other unrelated issues that require simple schema changes, and get ready for a v24 release.

inventor02 commented:
Hey - sorry it's been a while, I've not had much free time recently! Feel free to take/update/merge - let me know if you'd like me to do anything/have any questions.

sde1000 commented:
Here are the changes I made after merging the pull request: https://github.com/sde1000/quicktill/commit/bc06249a578ad81063b8f664b4f3c3c15d9dc0a8

The changes from v23.15 to that commit (i.e. the whole feature as introduced) are here: https://github.com/sde1000/quicktill/compare/v23.15...bc06249a578ad81063b8f664b4f3c3c15d9dc0a8

Next time I get a chance to look at this, I'll add the remaining features I think are needed: avoiding the need for the user to press Cash/Enter on password entry, a plugin system for different password hash algorithms, and support for changing passwords through the web interface.

Thank you very much for your contribution! I don't think I'd have got around to implementing it nearly as quickly without your help.
